pkg/virt-controller/watch/migration.go

```diff
@@ -431,10 +474,143 @@ func (c *MigrationController) createTargetPod(migration *virtv1.VirtualMachineIn
+		return err
+	}
+	log.Log.Object(vmi).Infof("Handed off migration %s/%s to target virt-handler.", migration.Namespace, migration.Name)
+	c.recorder.Eventf(migration, k8sv1.EventTypeNormal, SuccessfulHandOverPodReason, "Migration target pod is ready for preparation by virt-handler.")
+	return nil
+}
+
+func (c *MigrationController) handleSignalMigrationAbort(migration *virtv1.VirtualMachineInstanceMigration, vmi *virtv1.VirtualMachineInstance) error {
+
+	vmiCopy := vmi.DeepCopy()
+
+	vmiCopy.Status.MigrationState.AbortRequested = true
+	if !reflect.DeepEqual(vmi.Status, vmiCopy.Status) {
+		newStatus, err := json.Marshal(vmiCopy.Status)
+		if err != nil {
+			return err
+		}
+		oldStatus, err := json.Marshal(vmi.Status)
+		if err != nil {
+			return err
+		}
+		test := fmt.Sprintf(`{ "op": "test", "path": "/status", "value": %s }`, string(oldStatus))
+		patch := fmt.Sprintf(`{ "op": "replace", "path": "/status", "value": %s }`, string(newStatus))
+		_, err = c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(vmi.Name, types.JSONPatchType, []byte(fmt.Sprintf("[ %s, %s ]", test, patch)))
+		if err != nil {
+			msg := fmt.Sprintf("failed to set MigrationState in VMI status. :%v", err)
+			c.recorder.Eventf(migration, k8sv1.EventTypeWarning, FailedAbortMigrationReason, msg)
+			return fmt.Errorf(msg)
+		}
+		log.Log.Object(vmi).Infof("Signaled migration %s/%s to be aborted.", migration.Namespace, migration.Name)
+		c.recorder.Eventf(migration, k8sv1.EventTypeNormal, SuccessfulAbortMigrationReason, "Migration is ready to be canceled by virt-handler.")
+	}
+
+	return nil
+}
+
+func (c *MigrationController) handleTargetPodCreation(key string, migration *virtv1.VirtualMachineInstanceMigration, vmi *virtv1.VirtualMachineInstance) error {
+
+	c.migrationStartLock.Lock()
+	defer c.migrationStartLock.Unlock()
+
+	// Don't start new migrations if we wait for cache updates on migration target pods
+	if c.podExpectations.AllPendingCreations() > 0 {
+		c.Queue.AddAfter(key, 1*time.Second)
+		return nil
+	} else if controller.VMIActivePodsCount(vmi, c.podInformer) > 1 {
+		log.Log.Object(migration).Infof("Waiting to schedule target pod for migration because there are already multiple pods running for vmi %s/%s", vmi.Namespace, vmi.Name)
+		c.Queue.AddAfter(key, 1*time.Second)
+		return nil
+
+	}
+
+	// Don't start new migrations if we wait for migration object updates because of new target pods
+	runningMigrations, err := c.findRunningMigrations()
+	if err != nil {
+		return fmt.Errorf("failed to determin the number of running migrations: %v", err)
+	}
+
+	// XXX: Make this configurable, think about limit per node, bandwidth per migration, and so on.
+	if len(runningMigrations) >= int(*c.clusterConfig.GetMigrationConfiguration().ParallelMigrationsPerCluster) {
+		// Let's wait until some migrations are done
+		c.Queue.AddAfter(key, time.Second*5)
```

Comment:
Notwithstanding the note about this being configurable:

Would it make sense to add some randomness here? e.g. instead of a hard 5 seconds, a random interval between 5 and 10?

Reply:
i'm not completely sure actually. i can't come up with a reason for why adding jitter here would help. a backoff might help a bit if lots of migrations are being worked on (probably rare). In this case, the static delay is probably a fairly pragmatic solution though.